src/sympc/tensor/mpc_tensor.py

```diff
@@ -373,43 +371,22 @@ def reconstruct(
         Returns:
             torch.Tensor. The secret reconstructed.
         """
+        plaintext = self.session.protocol.share_class.reconstruct(
+            self.share_ptrs, get_shares
+        )
 
-        def _request_and_get(share_ptr: ShareTensor) -> ShareTensor:
-            """Function used to request and get a share - Duet Setup.
-
-            Args:
-                share_ptr (ShareTensor): a ShareTensor
-
-            Returns:
-                ShareTensor. The ShareTensor in local.
-
-            """
-            if not islocal(share_ptr):
-                share_ptr.request(block=True)
-            res = share_ptr.get_copy()
-            return res
-
-        request = _request_and_get
-        request_wrap = parallel_execution(request)
-
-        args = [[share] for share in self.share_ptrs]
-        local_shares = request_wrap(args)
-
-        shares = [share.tensor for share in local_shares]
-
-        if get_shares:
-            return shares
-
-        plaintext = sum(shares)
-
-        if decode:
+        if decode and not get_shares:
             fp_encoder = FixedPointEncoder(
                 base=self.session.config.encoder_base,
                 precision=self.session.config.encoder_precision,
             )
 
             plaintext = fp_encoder.decode(plaintext)
 
+        else:
+
+            return plaintext
```

Comment:
MAYBE : Currently we do not allow if `decode = True` and `get_shares=True` , we only output the shares without decoding even if the parameter is true. Should we allow it?I am not sure of a use case where this would be useful,we might have in the future.

Reply:
Hmmmm, this was present in the original implementation of sharetensor. Not sure why it was intended. That's a good question. @gmuraru Why does the orchestrator need the original shares instead of just value?

Reply:
The `get_shares` and `decode` it is used in the spdz protocol.
Maybe we should add a disclaimer to the function saying that ```get_shares``` has priority over `decode`.

In SPDZ `decode` is used when we reconstruct the `eps` and `delta` values because the MPC tensor has attached a session with config, but we do not intend to decode those values.
The `get_shares` --- hmmm. I think it could be removed. and in the SPDZ we could replace:
```
      z_shares_local = z.get_shares()
      common_args = [z_shares_local, y]
```
with
```
      common_args = [z.share_ptrs, y]
```

Could you try to see if it works?

Reply:
I think SPDZ doesn't use it anymore. https://github.com/OpenMined/SyMPC/blob/5144a24779a72778c22a2a539b2b5e1aca4f2991/src/sympc/protocol/spdz/spdz.py#L128.